core/http-auth-aws-crt/src/main/java/software/amazon/awssdk/http/auth/aws/crt/internal/signer/V4aContext.java

```diff
@@ -0,0 +1,52 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.http.auth.aws.crt.internal.signer;
+
+import software.amazon.awssdk.annotations.SdkInternalApi;
+import software.amazon.awssdk.crt.auth.signing.AwsSigningConfig;
+import software.amazon.awssdk.crt.http.HttpRequest;
+import software.amazon.awssdk.http.SdkHttpRequest;
+
+/**
+ * A container for data produced during and as a result of the SigV4a request signing with CRT.
+ */
+@SdkInternalApi
+public final class V4aContext {
+    private final SdkHttpRequest signedRequest;
+    private final HttpRequest signedCrtRequest;
+    private final byte[] signature;
+    private final AwsSigningConfig signingConfig;
+
+    public V4aContext(SdkHttpRequest signedRequest, HttpRequest signedCrtRequest, byte[] signature,
+                      AwsSigningConfig signingConfig) {
+        this.signedRequest = signedRequest;
+        this.signedCrtRequest = signedCrtRequest;
+        this.signature = signature.clone();
+        this.signingConfig = signingConfig;
+    }
+
+    public SdkHttpRequest getSignedRequest() {
+        return signedRequest;
+    }
+
+    public byte[] getSignature() {
+        return signature.clone();
```

Comment:
Yes, it's reported as a bug if not (returning a mutable byte array).

Reply:
Feels like unnecessary clone. Maybe we can add suppression? https://github.com/aws/aws-sdk-java-v2/blob/master/build-tools/src/main/resources/software/amazon/awssdk/spotbugs-suppressions.xml#L135

Reply:
In general, it makes sense to make defensive clone before returning mutable array. If the point is that this is SdkInternal and we trust the callers and how it is used, that may be ok, but if signatures are not expensive to clone, maybe cloning is fine?

Reply:
Yeah, I agree we should make a copy and we actually do in the ctor already. I'm just not sure if we should clone it again in the getter method.

Reply:
I was talking to the point of cloning the array 2nd time. Cloning in the constructor is to prevent the client that instantiated the object from mutating this object's state. Cloning in the accessor is to prevent the client that receives the array reference from mutating this object's state. If this was a SdkPublicApi class, this would be necessary. For SdkInternalApi, we _could_ avoid if you think so.

Reply:
FWIW, I found [this existing class](https://github.com/aws/aws-sdk-java-v2/blob/master/core/auth/src/main/java/software/amazon/awssdk/auth/signer/internal/SignerKey.java) that clones twice like that, incidentally related to signing too.

Reply:
Yeah, it's not consistent in the SDK for now. IMO this is a bit overkill for internal APIs, but it's definitely open for discussion.